First pass of docstring additions.

nsnw opened the pull request:
Hey! With reference to #1565, I've started tidying up some of the docstrings in `bot.py`. I'm happy to help do others as well, but this is just to check that this is the kind of thing you're looking for. Feedback is obviously requested and will be gladly received!

dgw commented on sopel/bot.py:
```diff
@@ -174,10 +183,16 @@ def write(self, args, text=None):  # Shim this in here for autodocs
         sending. Additionally, if the message (after joining) is longer than
         than 510 characters, any remaining characters will not be sent.
+
+        :param collections.Iterable args: an iterable of strings, which will be joined by spaces
```
You'd need to use `:type:` lines also, but it's possible to link things like this to the Python glossary:

```
:param args: command arguments, which will be joined by spaces
:type args: :term:`iterable`
```

That produces the type hint "([iterable](https://docs.python.org/3/glossary.html#term-iterable))" for `args`, linking to the term in Python's own docs. Feel free to use this trick all over the place! There are many references to such things in this PR.

nsnw replied:
Thanks for this hint - I'm reasonably new to Sphinx so this is a learning experience. I've used `:term:` in a couple of places in the latest revision!

dgw commented:
I had all but given up trying to figure out how to link `re.Pattern` as an example, so I left it out of my review.

But! This worked when I tweaked the current PR state locally and built the HTML docs:

```
    def register_url_callback(self, pattern, callback):
        """Register a ``callback`` for URLs matching the regex ``pattern``

        :param pattern: compiled regex pattern to register
        :type pattern: :ref:`re.Pattern <python:re-objects>`
        :param function callback: callable object to handle matching URLs
[... etc ...]
```

`pattern`'s type then links to https://docs.python.org/3/library/re.html#re-objects by doing this, which is precisely the section needed. Why it's not linkable as `#re.Pattern` despite describing what appear to be objects of that type, I don't know. But this is _a_ solution to the "More intersphinx links" point I requested.

nsnw commented:
Thanks for the feedback! I'll make a start on incorporating the feedback :-D

nsnw commented on sopel/bot.py:
```diff
@@ -25,7 +25,6 @@
 from sopel.logger import get_logger
 import sopel.loader
 
-
```
Not actually sure how I ended up deleting that line - I've put it back in the latest revision!

nsnw commented on sopel/bot.py:
```diff
@@ -463,6 +519,11 @@ def call(self, func, sopel, trigger):
 
     def dispatch(self, pretrigger):
+        """Match an incoming, parsed message from the server and dispatch it to any registered
+        callables.
```
Yeah, it was clumsy wording - I've updated it with your suggestion.

nsnw commented on sopel/bot.py:
```diff
@@ -263,7 +291,11 @@ def register(self, callables, jobs, shutdowns, urls):
+        """Part a channel.
+
+        :param str channel: the channel to leave
+        :param str msg: the message to display when leaving a channel
```
I settled on **leave** (even though the method is `part`) as it read better - updated.

nsnw commented on sopel/bot.py:
```diff
@@ -237,6 +256,15 @@ def unregister(self, obj):
+        :param collections.Iterable urls: an iterable of functions to call when matched against a
+            URL
+        """
+
```
Noted!

nsnw commented:
I've incorporated the feedback, and:-

* docstrings are now limited to 80 characters wide
* moved parameter definitions to the top of the docstrings (under the initial one-line description)
* switching to hanging indentation for line-wrapped parameter definitions
* added glossary links for concepts (using `:term:` as suggested)
* removed blank newlines between docstring definitions and method bodies

dgw commented on sopel/bot.py:
```diff
@@ -438,12 +447,15 @@ def reply(self, text, dest, reply_to, notice=False):
-        :param Trigger trigger: the Trigger object for the line from the server that triggered
-            this call
+        """Call a function (``func``), after checking if the function is
+        restricted or subject to limits.
```
As above, no need to point out the parameter name in the short description when we have param lists.

Would also really like to keep short descriptions to one line only. Can't make suggestions that apply to multiple lines, so apply this (and tweak it, if you have a better idea) however you may manually:

```
    """Call a function, applying any rate-limits or restrictions.
```

nsnw replied:
Done!

dgw commented on sopel/bot.py:
```diff
@@ -519,8 +531,7 @@ def call(self, func, sopel, trigger):
     def dispatch(self, pretrigger):
-        """Match an incoming, parsed message from the server and dispatch it to any registered
-        callables.
+        """Dispatch a parsed message to any registered callables
```
I think we're trying to be consistent about having `.` at the end of these short ones? Maybe?

Even if we aren't yet, we probably should be.

nsnw replied:
That one slipped through - added.

nsnw commented on sopel/bot.py:
```diff
@@ -387,35 +392,44 @@ def say(self, text, recipient, max_messages=1):
-        the channel (or nickname, if a private message), in which the trigger
-        happened.
+        :param str text: the text to send in the NOTICE
+        :param str dest: the destination of the NOTICE
```
Whoops - missed that one when doing a local merge!

nsnw commented on sopel/bot.py:
```diff
@@ -387,35 +392,44 @@ def say(self, text, recipient, max_messages=1):
 
     def notice(self, text, dest):
-        """Send an IRC NOTICE to a user or a channel.
+        """Send an IRC NOTICE (of ``text``) to a user or a channel (``dest``).
```
Updated.

nsnw commented on sopel/bot.py:
```diff
@@ -387,35 +392,44 @@ def say(self, text, recipient, max_messages=1):
+        :param str dest: the destination of the reply
+        :param str reply_to: the nickname that the reply will be prepended with
+        :param bool notice: whether to send the reply as a NOTICE or not,
+                            defaults to False
```
Done!

nsnw commented on sopel/bot.py:
```diff
@@ -387,35 +392,44 @@ def say(self, text, recipient, max_messages=1):
         self.say('\001ACTION {}\001'.format(text), dest)
 
     def reply(self, text, dest, reply_to, notice=False):
         """Prepend ``reply_to`` to ``text``, and send as a PRIVMSG to ``dest``.
```
Assuming you're referring to 407-409, I've tried to think of an alternative way to write it and I can't think of anything that is that much shorter. Alternatively, maybe it's better to not mention it at all?

dgw replied:
I actually do mean this specific line. Let me take a stab at it, to give you at least a starting point maybe?

```suggestion
        """Send a PRIVMSG addressing someone.
```

Could end this with "to a user or channel" like the above `action` one, but probably not needed. In fact, I think I'll suggest removing most of those.

Might have shortened it a bit _too_ much, but it does eliminate naming the params… Hmm.

dgw commented on sopel/bot.py:
```diff
@@ -372,6 +421,12 @@ def action(self, text, dest):
+        :param str dest: the destination of the reply
+        :param str reply_to: the nickname that the reply will be prepended with
+        :param bool notice: whether to send the reply as a NOTICE or not,
+                            defaults to `False`
```
We need the double backticks, actually. Single backticks produce the wrong output (usually italics).

```suggestion
                            defaults to ``False``
```

dgw commented on sopel/bot.py:
```diff
@@ -351,14 +394,20 @@ def say(self, text, recipient, max_messages=1):
 
     def action(self, text, dest):
-        """Send ``text`` as a CTCP ACTION PRIVMSG to ``dest``.
+        """Send a CTCP ACTION PRIVMSG to a user or a channel.
```
```suggestion
        """Send a CTCP ACTION PRIVMSG to a user or channel.
```

Nitpicking, only because I already have other comments to make in this review, and I noticed inconsistencies.

dgw commented on sopel/bot.py:
```diff
@@ -351,14 +394,20 @@ def say(self, text, recipient, max_messages=1):
     def notice(self, text, dest):
         """Send an IRC NOTICE to a user or a channel.
```
Matching below.

```suggestion
        """Send an IRC NOTICE to a user or channel.
```

dgw commented on sopel/bot.py:
```diff
@@ -263,16 +295,22 @@ def register(self, callables, jobs, shutdowns, urls):
         self.write(['PART', channel], msg)
 
     def join(self, channel, password=None):
         """Join a channel
```
```suggestion
        """Join a channel.
```

Didn't notice the missing `.` before. :/

dgw commented on sopel/bot.py:
```diff
@@ -144,7 +151,10 @@ def __init__(self, config, daemon=False):
     @property
     def hostmask(self):
-        """str: the current hostmask for the bot :class:`sopel.tools.target.User`
+        """The current hostmask for the bot :class:`sopel.tools.target.User`
```
```suggestion
        """The current hostmask for the bot :class:`sopel.tools.target.User`.
```

dgw commented on sopel/bot.py:
```diff
@@ -286,6 +324,11 @@ def msg(self, recipient, text, max_messages=1):
     def say(self, text, recipient, max_messages=1):
         """Send ``text`` as a PRIVMSG to ``recipient``.
```
Overlooked this technical-soup one the other day. Sorry about that! Probably because you didn't touch it, and I was focused on the changed/added lines.

```suggestion
        """Send a PRIVMSG to a user or channel.
```

nsnw commented:
You're not being too demanding!

I've pushed the latest version. I've added the "to a user or channel" wording to all the ones that it makes for now. I personally can't decide which is better - with or without it - but I'll leave it to you to decide.

dgw commented on sopel/bot.py:
```diff
@@ -419,13 +419,14 @@ def action(self, text, dest):
 
     def reply(self, text, dest, reply_to, notice=False):
-        """Prepend ``reply_to`` to ``text``, and send as a PRIVMSG to ``dest``.
+        """Send a PRIVMSG to a user or channel, optionally prepended with a
```
Thing is, the text to prepend is _not_ optional. The argument has no default value, and is therefore required.

dgw commented:
> You're not being too demanding!

That's a relief! All the same, I thank you for putting up with the lengthy iteration process. :smile_cat:

> I've added the "to a user or channel" wording to all the ones that it makes for now. I personally can't decide which is better - with or without it - but I'll leave it to you to decide.

Obviously, I choose to keep "to a user or channel" in places where it makes sense to include, but without the second "a" that (to me) makes the phrase slightly redundant.

dgw commented on sopel/bot.py:
```diff
@@ -60,17 +60,21 @@ def __init__(self, config, daemon=False):
         declared. The first item in a callable's commands list is used as the
         key in version *3.2* onward. Prior to *3.2*, the name of the function
         as declared in the source code was used.
         """
```
This docstring could be rewritten to use Sphinx notations like `.. versionchanged:: 3.2`, with the notes below that. For example:

```py
        """A dictionary of command names to their documentation.

        Each command is mapped to its docstring and any available examples, if
        declared in the module's code.

        .. versionchanged:: 3.2
            Use the first item in each callable's commands list as the key,
            instead of the function name as declared in the source code.
        """
```

dgw commented on sopel/bot.py:
```diff
@@ -110,6 +116,7 @@ def __init__(self, config, daemon=False):
         :class:`sopel.tools.target.Channel` objects which contain the users in
         the channel and their permissions.
         """
+
```
Sigh, more out-of-reach lines.

Both above and below here, there are mentions of `Identifiers` that could be linked to <tt>:class:\`sopel.tools.Identifier\`</tt>

Below, there are also two docstrings (for `memory` and `shutdown_methods`) that should have trailing `.` added.

dgw commented on sopel/bot.py:
```diff
@@ -677,6 +759,7 @@ def unregister_url_callback(self, pattern):
         """Unregister the callback for URLs matching the regex ``pattern``
```
```suggestion
        """Unregister the callback for URLs matching the regex ``pattern``.
```

dgw commented on sopel/bot.py:
```diff
@@ -646,7 +726,9 @@ def register_url_callback(self, pattern, callback):
         """Register a ``callback`` for URLs matching the regex ``pattern``
```
```suggestion
        """Register a ``callback`` for URLs matching the regex ``pattern``.
```

dgw commented on sopel/bot.py:
```diff
@@ -263,16 +295,22 @@ def register(self, callables, jobs, shutdowns, urls):
+        :param str channel: the channel to join
+        :param str password: an optional channel password
 
         If `channel` contains a space, and no `password` is given, the space is
```
Figured it out. Single-backtick gives us `<cite>text</cite>`, which is grossly incorrect for these uses both semantically and visually (it renders in italicized prose font). Using double-backticks yields (I'm simplifying) `<code>text</code>`, which is what we want (renders as monospace).

```suggestion
        If ``channel`` contains a space, and no ``password`` is given, the space is
```

dgw commented on sopel/bot.py:
```diff
@@ -263,16 +295,22 @@ def register(self, callables, jobs, shutdowns, urls):
 
         If `channel` contains a space, and no `password` is given, the space is
         assumed to split the argument into the channel to join and its
         password.  `channel` should not contain a space if `password` is given.
```
```suggestion
        password. ``channel`` should not contain a space if ``password`` is given.
```

dgw commented on sopel/bot.py:
```diff
@@ -578,8 +647,19 @@ def cap_req(self, module_name, capability, arg=None, failure_callback=None,
+                                 capability is successfully requested
+        :type success_callback: :term:`function`
+
         By prefixing the capability with `-`, it will be ensured that the
```
Some of them are out of line-note range (GitHub is really pissing me off tonight with that limitation), but every instance of <tt>\`thing\`</tt> in this docstring needs to be replaced with <tt>\`\`thing\`\`</tt>. It's the only place where single-backticks are used (based on searching the generated HTML for incorrect occurrences of `<cite>`) other than the `join` method docstring.

I would also appreciate a general cleanup of any remaining double-spaces after periods, as occurs on what is currently line 671 of the HEAD revision (but that no doubt will have changed by the time you address this note :sweat_smile:)!

nsnw commented:
I am still working on this! I've just been busy with work. Hoping to get this finished in the next few days.

Exirel commented:
Just a quick note to say: I love it! :+1: 
Also you'll have to rebase and resolve some conflict. And the more PR get merged, the more conflicts you'll have to fix.

nsnw commented:
Ok! Sorry for the delay - currently in the process of buying a house so things got in the way -_-

In the latest commit I've implemented the suggestions, including:-
* switching single backticks to double backticks where appropriate - a scan of the generated HTML after this doesn't show any instances of `<cite>`/`</cite>` 
* added periods at the end of sentences - I think I've now caught them all
* removed 2 instances of double spaces after a period

There were a couple of other bits I've found as well that I fixed.

nsnw commented:
I've merged in the latest changes from `master`, so this should be current against those.

dgw commented on sopel/bot.py:
```diff
@@ -186,9 +191,10 @@ def write(self, args, text=None):  # Shim this in here for autodocs
+        Newlines and carriage returns (``'\\n'`` and ``'\\r'``) are
+        removed before sending. Additionally, if the message (after joining) is
+        longer than than 510 characters, any remaining characters will not be
+        sent.
```
Rewrapping this paragraph should be able to eliminate this widow.

dgw commented on sopel/bot.py:
```diff
@@ -66,10 +66,14 @@ def __init__(self, config, daemon=False):
-        declared. The first item in a callable's commands list is used as the
-        key in version *3.2* onward. Prior to *3.2*, the name of the function
-        as declared in the source code was used.
+        A dictionary of command names to their documentation.
```
This should go on the same line as `"""`.

nsnw commented:
Excellent! Been a bit busy (getting married in a few weeks!) but I'll make the tweaks above and tidy up the PR. Once that's done I'll start looking at others :-)

Exirel commented:
@nsnw that's lovely! :heart_eyes:

nsnw commented:
Okay - I believe I've squashed everything down to a single commit now. I have to admit that squashing and rebasing is not my strong point so if I've butchered it I apoplogise!
